File: misc/aooswd/ex/memento/memento2.java
public class MoveCommand extends Command
{
	private ConstraintSolverMemento _state;
	private Point _delta;
	private Graphic _target;

	public MoveCommand (Graphic target, Point Delta)
	{
		_delta = Delta;
		_target = target;
	}

	public void Execute ()
	{
		ConstraintSolver solver = new ConstraintSolver ();
		_state = solver.CreateMemento();
		_target.Move(_delta);
		solver.Solve();
	}

	public void Unexecute()
	{
		ConstraintSolver solver = new ConstraintSolver ();
		_target.Move (_delta);
		solver.SetMemento (_state);
		solver.Solve ();
	}
}

public class ConstraintSolver extends Object
{
	public ConstraintSolver ()
	{
	}

	public void Solve ();

	public void AddConstraint (Graphic startConnection, Graphic endConnection);

	public RemoveConstraint (Graphic startConnection, Graphic endConnection);

	public ConstraintSolverMemento CreateMemento ();

	public void SetMemento (ConstraintSolverMemento M);
}

File: misc/aooswd/ex/memento/memento2.cpp
class MoveCommand
{
public:
	MoveCommand (Graphic* target, const Point& Delta);
	void Execute ();
	void Unexecute ();
private:
	ConstraintSolverMemento* _state;
	Point _delta;
	Graphic* _target;
};

class ConstraintSolver
{
public:
	static ConstraintSolver* Instance ();
	void Solve ();
	void AddConstraint (Graphic* startConnection, Graphic* endConnection);
	void RemoveConstraint (Graphic* startConnection, Graphic* endConnection);
	ConstraintSolverMemento* CreateMemento ();
	void SetMemento (ConstraintSolverMemento*);
private:
…
};

class ConstraintSolverMemento
{
public:
	virtual ~ConstraintSolverMemento ();
private:
	friend class ConstraintSolver;
	ConstraintSolverMemento ();
};

void MoreCommand::Execute
{
	ConstraintSolver* solver = ConstraintSolver::Instance ();
	_state = solver -> CreateMemento ();
	_target -> Move (delta);
	solver -> Solve ();
}

void MoveCommand::Unexecute ()
{
	ConstraintSolver* solver = ConstraintSolver::Instance ();
	_target -> Move (-_delta);
	solver -> SetMemento (_state);
	solver -> Solve ();
}
